api/kubelb.k8c.io/v1alpha1/route_types.go

```diff
@@ -0,0 +1,140 @@
+	// +optional
+	Source RouteSource `json:"source,omitempty"`
+}
+
+type RouteSource struct {
+	// Kubernetes contains the information about the Kubernetes source.
+	Kubernetes *KubernetesSource `json:"kubernetes,omitempty"`
+}
+
+type KubernetesSource struct {
+	// TODO: Add more resources
+	// - gateway.networking.k8s.io/httproute
+	// - gateway.networking.k8s.io/grpcroute
+	// - gateway.networking.k8s.io/tcproute
+	// - gateway.networking.k8s.io/udproute
+
+	// Resources contains the list of resources that are used as the source for the Route.
+	// Allowed resources:
+	// - networking.k8s.io/ingress
+	// +optional
+	// +kubebuilder:validation:EmbeddedResource
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Route unstructured.Unstructured `json:"resource,omitempty"`
+
+	// Services contains the list of services that are used as the source for the Route.
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Services []UpstreamService `json:"services,omitempty"`
+
+	// ReferenceGrants contains the list of ReferenceGrants that are used as the source for the Route.
+	// ReferenceGrant identifies kinds of resources in other namespaces that are
+	// trusted to reference the specified kinds of resources in the same namespace
+	// as the policy.
+	// +kubebuilder:pruning:PreserveUnknownFields
+	ReferenceGrants []UpstreamReferenceGrant `json:"referenceGrants,omitempty"`
+}
+
+// TODO(waleed): Evaluate if this is really worth it, semantically it makes sense but it adds a lot of boilerplate. Alternatively,
+// we can simply use YQ and add the required markers/fields in the CRD. This will also make the CRD more readable.
+
+// UpstreamService is a wrapper over the corev1.Service object.
+// This is required as kubebuilder:validation:EmbeddedResource marker adds the x-kubernetes-embedded-resource to the array instead of
+// the elements within it. Which results in a broken CRD; validation error. Without this marker, the embedded resource is not properly
+// serialized to the CRD.
+type UpstreamService struct {
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +kubebuilder:validation:EmbeddedResource
+	corev1.Service `json:",inline"`
+}
+
+// UpstreamReferenceGrant is a wrapper over the sigs.k8s.io/gateway-api/apis/v1alpha2.ReferenceGrant object.
+// This is required as kubebuilder:validation:EmbeddedResource marker adds the x-kubernetes-embedded-resource to the array instead of
+// the elements within it. Which results in a broken CRD; validation error. Without this marker, the embedded resource is not properly
+// serialized to the CRD.
+type UpstreamReferenceGrant struct {
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +kubebuilder:validation:EmbeddedResource
+	gwapiv1a2.ReferenceGrant `json:",inline"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// Route is the object that represents a route in the cluster.
+type Route struct {
```

Comment:
For reviewer:

There will be another dedicated CRD for Gateway and it's reference grants